### PR TITLE
ENH: Add file buffer validation to I/O ops

### LINKED ISSUE #15337: Segfault when passing a mock.Mock instance to pandas.read_csv (it was an accident!)
#### Code Sample

```python
In [1]: import pandas as pd

In [2]: from unittest import mock

In [3]: pd.read_csv(mock.Mock())
zsh: segmentation fault  ipython

```
#### Problem description

I wrote a test where I accidentally ended up passing a `mock.Mock` object into `pandas.read_csv` and it segfaults.

#### Expected Output

Well, it would nice if it threw an error telling me that I screwed up, instead of segfaulting.

#### Output of ``pd.show_versions()``

<details>
In [6]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Darwin
OS-release: 16.3.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: None
pip: 9.0.1
setuptools: 24.0.2
Cython: 0.24.1
numpy: 1.10.4
scipy: None
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.9
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext)
jinja2: 2.8
boto: None

</details>

### PR DESCRIPTION
**Pull request by gfyoung**
1) Allows for more uniform handling of invalid file buffers to our `read_*` functions.
2) Adds a ton of new documentation to `inference.py`

Closes #15337.

Partially addresses #15895.

### CODE REVIEW

File: pandas/types/inference.py [hunk 1]
```diff
@@ -35,8 +35,28 @@ def _iterable_not_string(x):
...
+    if PY2:
+        return hasattr(obj, 'next')
+    else:
+        # Python 3 generators have
+        # __next__ instead of next
+        return hasattr(obj, '__next__')
+
+
+def is_file_like(obj):
+    file_attrs = ('read', 'write', 'seek', 'tell')
```
Comment (jreback) [inline pandas/types/inference.py]: can you add a doc-string describing this. (and of course any others you want to document; prob should open an issue for that).

File: pandas/types/inference.py [hunk 2]
```diff
@@ -35,8 +35,28 @@ def _iterable_not_string(x):
...
+
+    for attr in file_attrs:
+        if not hasattr(obj, attr):
+            return False
+
+    if not is_iterator(obj):
```
Comment (jreback) [inline pandas/types/inference.py]: so a file_like *must* be an iterator, but an iterator is not necessarily file_like right?

File: pandas/types/inference.py (hunk 1 above)
Comment (gfyoung) [inline pandas/types/inference.py]: Isn't too hard to fix here in this PR.  Will do.

File: pandas/types/inference.py (hunk 2 above)
Comment (gfyoung) [inline pandas/types/inference.py]: Correct.  The reason is the following:
~~~
>>> from unittest import mock
>>> m = mock.Mock()
>>> hasattr(m, "read")
True
>>> hasattr(m, "__iter__")
False
~~~

**Comment by jreback** [comment]
lgtm.

File: pandas/types/inference.py (hunk 1 above)
Comment (jreback) [inline pandas/types/inference.py]: also add a versionadded (for this one).

Comment (gfyoung) [inline pandas/types/inference.py]: Got it.  Will make sure.

File: pandas/types/inference.py (hunk 2 above)
Comment (jreback) [inline pandas/types/inference.py]: k, maybe add that statement (or describe). Remember these are actually user-level accessible :>

Comment (gfyoung) [inline pandas/types/inference.py]: That is true.  In the doc-string, I presume?

Comment (jreback) [inline pandas/types/inference.py]: yep exactly. after this is merged I am going to create an issue to improve these doc-strings (novice level).

Comment (gfyoung) [inline pandas/types/inference.py]: That won't be necessary.  A commit is being added here for that.

Comment (jreback) [inline pandas/types/inference.py]: https://github.com/pandas-dev/pandas/issues/15895

File: pandas/io/excel.py [hunk 3]
```diff
@@ -241,11 +241,15 @@ def __init__(self, io, **kwds):
 
         # If io is a url, want to keep the data as bytes so can't pass
         # to get_filepath_or_buffer()
-        if _is_url(io):
+        url_passed = _is_url(io)
+
+        if url_passed:
             io = _urlopen(io)
```
Comment (jreback) [inline pandas/io/excel.py]: might be simpler to have this in an if/elif

```
if _is_url(io):
    io = _urlopen(io)
elif not isinstance(io, .....)
   ...
```

File: pandas/types/inference.py (hunk 2 above)
Comment (jreback) [inline pandas/types/inference.py]: yeah should document them all ...

Comment (jreback) [inline pandas/types/inference.py]: examples too!

File: pandas/io/excel.py (hunk 3 above)
Comment (gfyoung) [inline pandas/io/excel.py]: Good point.  Fixed.

**Comment by gfyoung** [comment]
@jreback : Everything is green and ready to go.

File: pandas/types/inference.py [hunk 4]
```diff
@@ -22,28 +22,209 @@
...
+
+    return (isinstance(obj, collections.Iterable) and
+            not isinstance(obj, string_types))
 
 
 def is_iterator(obj):
-    # python 3 generators have __next__ instead of next
-    return hasattr(obj, 'next') or hasattr(obj, '__next__')
+    """
+    Check if the object is an iterable.
```
Comment (jorisvandenbossche) [inline pandas/types/inference.py]: Maybe good to notice here that a string does not give True (although it is iterable, but not an iterator)

File: pandas/types/inference.py [hunk 5]
```diff
@@ -52,21 +233,86 @@ def is_re_compilable(obj):
         return True
 
 
-def is_list_like(arg):
-    return (hasattr(arg, '__iter__') and
-            not isinstance(arg, string_and_binary_types))
+def is_list_like(obj):
+    """
+    Check if the object is list-like.
```
Comment (jorisvandenbossche) [inline pandas/types/inference.py]: Maybe give an inexhaustive list of objects that are regarded as list-like? (lists, sets, arrays, Series, ...)

I would also add example with np.array(..)

**Review by jorisvandenbossche — COMMENTED** [review]
Nice docstrings!

File: pandas/types/inference.py (hunk 4 above)
Comment (jreback) [inline pandas/types/inference.py]: nor is a datetime, yes would be good to mention these counter-examples

File: pandas/types/inference.py (hunk 5 above)
Comment (jreback) [inline pandas/types/inference.py]: mention that strings & datetimes are NOT is_list_like .

Comment (gfyoung) [inline pandas/types/inference.py]: Done.

File: pandas/types/inference.py (hunk 4 above)
Comment (gfyoung) [inline pandas/types/inference.py]: Done.

File: pandas/types/inference.py [hunk 6]
```diff
@@ -22,28 +22,216 @@
...
+
+    For objects to be considered file-like, they must
+    have the following four methods:
+
+    1) read
+    2) write
+    3) seek
+    4) tell
+
+    In addition, the object must be iterable, as mock
```
Comment (jreback) [inline pandas/types/inference.py]: this is assuming a little bit of knowledge here. I would just eliminate this paragraph.

File: pandas/types/inference.py [hunk 7]
```diff
@@ -82,25 +344,48 @@ def is_hashable(arg):
...
+        hash(obj)
     except TypeError:
         return False
     else:
         return True
 
 
-def is_sequence(x):
+def is_sequence(obj):
+    """
```
Comment (jreback) [inline pandas/types/inference.py]: I would note that this is less strict than ``is_list_like``. Further impl wise we almost *never* use this one. I don't really recall what the actual diffferences are though. you can have a look thru the code base and see if you can come up with a good reason for this (IIRC its in the printing routines somewhere).

File: pandas/types/inference.py (hunk 6 above)
Comment (gfyoung) [inline pandas/types/inference.py]: Makes sense.  Done.

File: pandas/types/inference.py (hunk 7 above)
Comment (gfyoung) [inline pandas/types/inference.py]: Actually, I think it's the other way around:
~~~python
>>> i = iter([1, 2, 3])
>>> is_list_like(i)
True
>>> is_sequence(i)
False
~~~

Comment (gfyoung) [inline pandas/types/inference.py]: I think leaving both functions are fine for now, especially in the context of this PR.

**Comment by gfyoung** [comment]
@jreback , @jorisvandenbossche : Everything is green and ready to go.

File: pandas/types/inference.py (hunk 7 above)
Comment (jreback) [inline pandas/types/inference.py]: yes, just ideally wanted to document this.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**